contracts/verifying/VerifyingSingletonPaymasterV2.sol

```diff
@@ -1,9 +1,358 @@
+        if (sigLength != 65) revert InvalidPaymasterSignatureLength(sigLength);
+        //don't revert on signature failure: return SIG_VALIDATION_FAILED
+        if (
+            verifyingSigner != hash.toEthSignedMessageHash().recover(signature)
+        ) {
+            // empty context and sigFailed with time range provided
+            return (context, _packValidationData(true, validUntil, validAfter));
+        }
+
+        require(priceMarkup <= 2e6, "Verifying PM:high markup %");
+        // Review if below is needed (as paymaster service may pass 0 if dynamic pricing doesn't apply)
+        require(priceMarkup >= 1e6, "Verifying PM:low markup %");
+
+        // Review: may not be needed at all
+        address account = userOp.getSender();
+
+        // Review max or min
+        uint32 dynamicMarkup = maxuint32(priceMarkup, fixedPriceMarkup);
+
+        uint256 effectiveCost = (requiredPreFund * dynamicMarkup) /
+            PRICE_DENOMINATOR;
+
+        if (effectiveCost > paymasterIdBalances[paymasterId])
+            revert InsufficientBalance(
+                effectiveCost,
+                paymasterIdBalances[paymasterId]
+            );
+
+        context = abi.encode(
+            account,
+            paymasterId,
+            priceMarkup,
+            userOp.maxFeePerGas,
+            userOp.maxPriorityFeePerGas,
+            userOpHash
+        );
+
+        return (context, _packValidationData(false, validUntil, validAfter));
+    }
+
+    function parsePaymasterAndData(
+        bytes calldata paymasterAndData
+    )
+        public
+        pure
+        returns (
+            address paymasterId,
+            uint48 validUntil,
+            uint48 validAfter,
+            uint32 priceMarkup,
+            bytes calldata signature
+        )
+    {
+        (paymasterId, validUntil, validAfter, priceMarkup) = abi.decode(
+            paymasterAndData[VALID_PND_OFFSET:SIGNATURE_OFFSET],
+            (address, uint48, uint48, uint32)
+        );
+
+        signature = paymasterAndData[SIGNATURE_OFFSET:];
+    }
+
+    /**
+     * @dev Executes the paymaster's payment conditions
+     * @param mode tells whether the op succeeded, reverted, or if the op succeeded but cause the postOp to revert
+     * @param context payment conditions signed by the paymaster in `validatePaymasterUserOp`
+     * @param actualGasCost amount to be paid to the entry point in wei
+     */
+    function _postOp(
+        PostOpMode mode,
+        bytes calldata context,
+        uint256 actualGasCost
+    ) internal virtual override {
+        // Review what is needed in context
+        (
+            address account,
+            address paymasterId,
+            uint32 priceMarkup,
+            uint256 maxFeePerGas,
+            uint256 maxPriorityFeePerGas,
+            bytes32 userOpHash
+        ) = abi.decode(
```

Comment:
can remove account and userOpHash from context if they are not used in this method.

Reply:
hmm can do. as other events emitted in the transaction we can extract account and userOpHash info if needed (from events pov)